### ISSUE TITLE
User-defined patterns

### ISSUE THREAD

**Issue by gvanrossum** [gvanrossum/patma#13]
Since pattern matching is runtime behavior anyway, I wonder if we could have a way for users to construct their own patterns (without having to use `exec()`). For example:
```
P1 = ConstantPattern(int(input())
x = int(input())
match x:
    case P1: print("A match!")
    case _: print("Alas")
```
There could be various constructors for pattern objects that could be dynamically created and combined.

I'm not sure how to handle value extractions -- maybe `case P1(x)` would assign an extracted value to `x`? The value extracted would be up to the object `P1`.

But now it looks just like Object matching (`case Point(x, y, 0)` etc.). So maybe there's nothing we have to do? A pattern object would just be a class instance with an appropriate `__match__` method. The first example would have to be changed to `case P1():`.

**Comment by gvanrossum** [comment]
Something like this has come up in https://mail.python.org/archives/list/python-dev@python.org/message/DDYISJOZPJOD2C2RY4HH2XS3YMHBVYWX/

**Comment by Tobias-Kohn** [comment]
Yeah, I saw the post.  As far as I understood it, Federico Salerno's point of this is to solve the lead/store issue.  To that end, he is basically proposing a somewhat cumbersome implementation of proxy objects with quasi-scopes, rather than binding local variables as in the current PEP.  To be honest, I feel that if we really boil down his proposal in that post, most of it just evaporates and leads to basically the same discussions as we already had.

**Comment by gvanrossum** [comment]
Yeah, his proposal is the same as my poorly conceived idea here. But I feel it still needs some kind of rational response, since it's bringing up a *new* idea (for the list) rather than rehashing something old (like different sigils or keywords to mark loads or stores).

Though maybe we needn't respond until at least one other person supports his idea (this has roughly been my approach).
